File: day10/main.ts
import { readFileSync } from "fs";

type Machine = { pattern: string; buttons: number[][]; targets: number[] };

function parseMachine(line: string): Machine {
  const pat = /\[(.*?)\]/.exec(line)![1];
  const buttonMatches = [...line.matchAll(/\(([^)]*)\)/g)];
  const buttons = buttonMatches.map((m) =>
    m[1].trim()
      ? m[1]
          .split(",")
          .map((s) => parseInt(s.trim(), 10))
      : [],
  );
  const targets = /\{([^}]*)\}/
    .exec(line)![1]
    .split(",")
    .map((s) => parseInt(s.trim(), 10));
  return { pattern: pat, buttons, targets };
}

function loadMachines(path: string): Machine[] {
  return readFileSync(path, "utf8")
    .trim()
    .split(/\r?\n/)
    .filter(Boolean)
    .map(parseMachine);
}

function part1(ms: Machine[]): number {
  let total = 0;
  for (const m of ms) {
    const lights = m.pattern.length;
    let target = 0;
    for (let i = 0; i < lights; i++) {
      if (m.pattern[i] === "#") target |= 1 << i;
    }
    const masks = m.buttons.map((b) => {
      let mask = 0;
      for (const idx of b) {
        if (idx < lights) mask ^= 1 << idx;
      }
      return mask;
    });
    const n = masks.length;
    let best = Number.POSITIVE_INFINITY;
    for (let mask = 0; mask < 1 << n; mask++) {
      let state = 0;
      let presses = 0;
      for (let i = 0; i < n; i++) {
        if ((mask >> i) & 1) {
          state ^= masks[i];
          presses++;
        }
      }
      if (state === target && presses < best) best = presses;
    }
    total += best;
  }
  return total;
}

function part2(ms: Machine[]): number {
  let total = 0;
  const EPS = 1e-9;

  for (const m of ms) {
    const counters = m.targets.length;
    const buttons = m.buttons.length;
    if (counters === 0 || buttons === 0) continue;

    // Matrix counters x buttons
    const mat: number[][] = Array.from({ length: counters }, () =>
      Array(buttons).fill(0),
    );
    m.buttons.forEach((b, cidx) => {
      for (const t of b) {
        if (t >= 0 && t < counters) mat[t][cidx] = 1;
      }
    });
    const rhs = m.targets.map((v) => v * 1.0);

    // RREF
    const pivotCols = Array(counters).fill(-1);
    let row = 0;
    for (let col = 0; col < buttons && row < counters; col++) {
      let pivot = -1;
      let bestVal = 0;
      for (let r = row; r < counters; r++) {
        const val = Math.abs(mat[r][col]);
        if (val > EPS && val > bestVal) {
          bestVal = val;
          pivot = r;
        }
      }
      if (pivot === -1) continue;
      if (pivot !== row) {
        [mat[row], mat[pivot]] = [mat[pivot], mat[row]];
        [rhs[row], rhs[pivot]] = [rhs[pivot], rhs[row]];
      }
      const piv = mat[row][col];
      for (let c = col; c < buttons; c++) mat[row][c] /= piv;
      rhs[row] /= piv;
      for (let r = 0; r < counters; r++) {
        if (r === row) continue;
        const f = mat[r][col];
        if (Math.abs(f) < EPS) continue;
        for (let c = col; c < buttons; c++) mat[r][c] -= f * mat[row][c];
        rhs[r] -= f * rhs[row];
      }
      pivotCols[row] = col;
      row++;
    }
    const rank = row;
    let inconsistent = false;
    for (let r = rank; r < counters; r++) {
      let maxv = 0;
      for (let c = 0; c < buttons; c++) maxv = Math.max(maxv, Math.abs(mat[r][c]));
      if (maxv < EPS && Math.abs(rhs[r]) > EPS) {
        inconsistent = true;
        break;
      }
    }
    if (inconsistent) continue;

    const used = Array(buttons).fill(false);
    for (let i = 0; i < rank; i++) if (pivotCols[i] >= 0) used[pivotCols[i]] = true;
    const freeCols = [];
    for (let c = 0; c < buttons; c++) if (!used[c]) freeCols.push(c);
    const freeCount = freeCols.length;

    const coef: number[][] = Array.from({ length: rank }, () =>
      Array(freeCount).fill(0),
    );
    for (let r = 0; r < rank; r++) {
      for (let f = 0; f < freeCount; f++) {
        coef[r][f] = mat[r][freeCols[f]];
      }
    }

    const sumTargets = m.targets.reduce((a, b) => a + b, 0);
    let best = sumTargets;
    const freeVals = Array(freeCount).fill(0);

    const evaluate = (cur: number) => {
      if (cur >= best) return;
      let sum = cur;
      for (let r = 0; r < rank; r++) {
        let v = rhs[r];
        for (let f = 0; f < freeCount; f++) v -= coef[r][f] * freeVals[f];
        if (v < -EPS) return;
        const iv = Math.round(v);
        if (Math.abs(iv - v) > EPS) return;
        sum += iv;
        if (sum >= best) return;
      }
      if (sum < best) best = sum;
    };

    evaluate(0);

    const quick = (idx: number, cur: number, cap: number) => {
      if (cur >= best) return;
      if (idx === freeCount) {
        evaluate(cur);
        return;
      }
      for (let v = 0; v <= cap; v++) {
        if (cur + v >= best) break;
        freeVals[idx] = v;
        quick(idx + 1, cur + v, cap);
      }
    };

    let seedCap = 400;
    if (best < seedCap) seedCap = best;
    if (freeCount > 0 && seedCap > 0) quick(0, 0, seedCap);

    const dfs = (idx: number, cur: number) => {
      if (cur >= best) return;
      if (idx === freeCount) {
        evaluate(cur);
        return;
      }
      const maxv = best - cur;
      for (let v = 0; v <= maxv; v++) {
        freeVals[idx] = v;
        dfs(idx + 1, cur + v);
      }
    };

    if (freeCount > 0) dfs(0, 0);
    else evaluate(0);

    total += best;
  }
  return total;
}

const machines = loadMachines("input.txt");
const t0 = performance.now();
const p1 = part1(machines);
const p2 = part2(machines);
const elapsed = performance.now() - t0;
console.log(
  `min_lights_presses=${p1} min_counter_presses=${p2} elapsed_ms=${elapsed.toFixed(
    3,
  )}`,
);

File: day10/main.rs
use regex::Regex;
use std::fs;
use std::time::Instant;

#[derive(Clone)]
struct Machine {
    pattern: String,
    buttons: Vec<Vec<usize>>,
    targets: Vec<i32>,
}

fn parse_machine(line: &str) -> Machine {
    let pat_re = Regex::new(r"\[(.*?)\]").unwrap();
    let button_re = Regex::new(r"\(([^)]*)\)").unwrap();
    let target_re = Regex::new(r"\{([^}]*)\}").unwrap();

    let pattern = pat_re.captures(line).unwrap()[1].to_string();
    let buttons = button_re
        .captures_iter(line)
        .map(|cap| {
            let s = cap[1].trim();
            if s.is_empty() {
                Vec::new()
            } else {
                s.split(',')
                    .map(|t| t.trim().parse::<usize>().unwrap())
                    .collect::<Vec<usize>>()
            }
        })
        .collect::<Vec<_>>();
    let targets = target_re.captures(line).unwrap()[1]
        .split(',')
        .map(|t| t.trim().parse::<i32>().unwrap())
        .collect::<Vec<i32>>();
    Machine {
        pattern,
        buttons,
        targets,
    }
}

fn load(path: &str) -> Vec<Machine> {
    fs::read_to_string(path)
        .expect("input")
        .lines()
        .filter(|l| !l.trim().is_empty())
        .map(parse_machine)
        .collect()
}

fn part1(machines: &[Machine]) -> i32 {
    let mut total = 0;
    for m in machines {
        let lights = m.pattern.len();
        let mut target = 0u32;
        for (i, ch) in m.pattern.as_bytes().iter().enumerate() {
            if *ch == b'#' {
                target |= 1 << i;
            }
        }
        let button_masks: Vec<u32> = m
            .buttons
            .iter()
            .map(|b| {
                let mut mask = 0u32;
                for &idx in b {
                    if idx < lights {
                        mask ^= 1 << idx;
                    }
                }
                mask
            })
            .collect();
        let n = button_masks.len();
        let mut best: Option<i32> = None;
        for mask in 0..(1u32 << n) {
            let mut state = 0u32;
            let presses = mask.count_ones() as i32;
            for i in 0..n {
                if (mask >> i) & 1 == 1 {
                    state ^= button_masks[i];
                }
            }
            if state == target {
                if best.is_none() || presses < best.unwrap() {
                    best = Some(presses);
                }
            }
        }
        total += best.unwrap_or(0);
    }
    total
}

fn part2(machines: &[Machine]) -> i32 {
    let mut total = 0;
    const EPS: f64 = 1e-9;

    for m in machines {
        let counters = m.targets.len();
        let buttons = m.buttons.len();
        if counters == 0 || buttons == 0 {
            continue;
        }

        // Build matrix counters x buttons
        let mut mat = vec![vec![0.0f64; buttons]; counters];
        for (cidx, b) in m.buttons.iter().enumerate() {
            for &t in b {
                if t < counters {
                    mat[t][cidx] = 1.0;
                }
            }
        }
        let mut rhs: Vec<f64> = m.targets.iter().map(|v| *v as f64).collect();

        // RREF
        let mut pivot_cols = vec![-1isize; counters];
        let mut row = 0;
        for col in 0..buttons {
            if row >= counters {
                break;
            }
            let mut pivot = -1;
            let mut best = 0.0;
            for r in row..counters {
                let val = mat[r][col].abs();
                if val > EPS && val > best {
                    best = val;
                    pivot = r as isize;
                }
            }
            if pivot == -1 {
                continue;
            }
            let p = pivot as usize;
            mat.swap(row, p);
            rhs.swap(row, p);
            let piv = mat[row][col];
            for c in col..buttons {
                mat[row][c] /= piv;
            }
            rhs[row] /= piv;
            for r in 0..counters {
                if r == row {
                    continue;
                }
                let f = mat[r][col];
                if f.abs() < EPS {
                    continue;
                }
                for c in col..buttons {
                    mat[r][c] -= f * mat[row][c];
                }
                rhs[r] -= f * rhs[row];
            }
            pivot_cols[row] = col as isize;
            row += 1;
        }
        let rank = row;
        let mut inconsistent = false;
        for r in rank..counters {
            let mut maxv = 0.0;
            for c in 0..buttons {
                let v = mat[r][c].abs();
                if v > maxv {
                    maxv = v;
                }
            }
            if maxv < EPS && rhs[r].abs() > EPS {
                inconsistent = true;
                break;
            }
        }
        if inconsistent {
            continue;
        }

        let mut used = vec![false; buttons];
        for r in 0..rank {
            if pivot_cols[r] >= 0 {
                used[pivot_cols[r] as usize] = true;
            }
        }
        let free_cols: Vec<usize> = (0..buttons).filter(|&c| !used[c]).collect();
        let free_count = free_cols.len();

        let mut coef = vec![vec![0.0f64; free_count]; rank];
        for r in 0..rank {
            for (fidx, &fc) in free_cols.iter().enumerate() {
                coef[r][fidx] = mat[r][fc];
            }
        }

        let sum_targets: i32 = m.targets.iter().sum();
        let mut best = sum_targets;
        let mut free_vals = vec![0i32; free_count];

        fn evaluate(
            cur: i32,
            rank: usize,
            free_count: usize,
            rhs: &[f64],
            coef: &[Vec<f64>],
            free_vals: &[i32],
            eps: f64,
            best: &mut i32,
        ) {
            if cur >= *best {
                return;
            }
            let mut sum = cur;
            for r in 0..rank {
                let mut v = rhs[r];
                for f in 0..free_count {
                    v -= coef[r][f] * free_vals[f] as f64;
                }
                if v < -eps {
                    return;
                }
                let iv = v.round() as i32;
                if (v - iv as f64).abs() > eps {
                    return;
                }
                sum += iv;
                if sum >= *best {
                    return;
                }
            }
            if sum < *best {
                *best = sum;
            }
        }

        evaluate(0, rank, free_count, &rhs, &coef, &free_vals, EPS, &mut best);

        fn quick(
            idx: usize,
            cur: i32,
            cap: i32,
            free_count: usize,
            free_vals: &mut [i32],
            rank: usize,
            rhs: &[f64],
            coef: &[Vec<f64>],
            eps: f64,
            best: &mut i32,
        ) {
            if cur >= *best {
                return;
            }
            if idx == free_count {
                evaluate(cur, rank, free_count, rhs, coef, free_vals, eps, best);
                return;
            }
            for v in 0..=cap {
                if cur + v >= *best {
                    break;
                }
                free_vals[idx] = v;
                quick(
                    idx + 1,
                    cur + v,
                    cap,
                    free_count,
                    free_vals,
                    rank,
                    rhs,
                    coef,
                    eps,
                    best,
                );
            }
        }

        let mut seed_cap = 400;
        if best < seed_cap {
            seed_cap = best;
        }
        if free_count > 0 && seed_cap > 0 {
            quick(
                0,
                0,
                seed_cap,
                free_count,
                &mut free_vals,
                rank,
                &rhs,
                &coef,
                EPS,
                &mut best,
            );
        }

        fn dfs(
            idx: usize,
            cur: i32,
            free_count: usize,
            free_vals: &mut [i32],
            rank: usize,
            rhs: &[f64],
            coef: &[Vec<f64>],
            eps: f64,
            best: &mut i32,
        ) {
            if cur >= *best {
                return;
            }
            if idx == free_count {
                evaluate(cur, rank, free_count, rhs, coef, free_vals, eps, best);
                return;
            }
            let maxv = *best - cur;
            for v in 0..=maxv {
                free_vals[idx] = v;
                dfs(
                    idx + 1,
                    cur + v,
                    free_count,
                    free_vals,
                    rank,
                    rhs,
                    coef,
                    eps,
                    best,
                );
            }
        }

        if free_count > 0 {
            dfs(
                0,
                0,
                free_count,
                &mut free_vals,
                rank,
                &rhs,
                &coef,
                EPS,
                &mut best,
            );
        } else {
            evaluate(0, rank, free_count, &rhs, &coef, &free_vals, EPS, &mut best);
        }

        total += best;
    }
    total
}

fn main() {
    let machines = load("input.txt");
    let start = Instant::now();
    let p1 = part1(&machines);
    let p2 = part2(&machines);
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    println!(
        "min_lights_presses={} min_counter_presses={} elapsed_ms={:.3}",
        p1, p2, elapsed
    );
}
